test/common/testlib.py

```diff
@@ -302,69 +391,51 @@ def switch_to_frame(self, name: str | None) -> None:
 
         :param name: frame name
         """
-        self.cdp.set_frame(name)
+        if name is None:
+            self.switch_to_top()
+        else:
+            asyncio.run_coroutine_threadsafe(self.driver.switch_to_frame(name), self.loop).result()
 
     def switch_to_top(self) -> None:
         """Switch to the main frame
 
         Switch to the main frame from for example an iframe.
         """
-        self.cdp.set_frame(None)
+        self.driver.switch_to_top()
+
+    def allow_download(self) -> None:
+        """Allow browser downloads"""
+        # this is only necessary for headless chromium
+        if self.browser == "chromium":
+            self.cdp_command("Page.setDownloadBehavior", behavior="allow", downloadPath=str(self.driver.download_dir))
 
     def upload_files(self, selector: str, files: Sequence[str]) -> None:
         """Upload a local file to the browser
 
         The selector should select the <input type="file"/> element.
         Files is a list of absolute paths to files which should be uploaded.
         """
-        r = self.cdp.invoke("Runtime.evaluate", expression='document.querySelector(%s)' % jsquote(selector))
-        objectId = r["result"]["objectId"]
-        self.cdp.invoke("DOM.setFileInputFiles", files=files, objectId=objectId)
-
-    def raise_cdp_exception(
-        self, func: str, arg: str, details: Mapping[str, Any], trailer: str | None = None
-    ) -> Never:
-        # unwrap a typical error string
-        if details.get("exception", {}).get("type") == "string":
-            msg = details["exception"]["value"]
-        elif details.get("text", None):
-            msg = details.get("text", None)
-        else:
-            msg = str(details)
-        if trailer:
-            msg += "\n" + trailer
-        raise Error("%s(%s): %s" % (func, arg, msg))
+        element = self.bidi("script.evaluate", expression=f"ph_find({jsquote(selector)})",
+                            awaitPromise=False, target={"context": self.driver.context})["result"]
+        self.bidi("input.setFiles", context=self.driver.context, element=element, files=files)
 
     def inject_js(self, code: str) -> None:
         """Execute JS code that does not return anything
 
         :param code: a string containing JavaScript code
         :type code: str
         """
-        self.cdp.invoke("Runtime.evaluate", expression=code, trace=code,
-                        silent=False, awaitPromise=True, returnByValue=False, no_trace=True)
+        # this is redundant now; the difference used to be important with CDP
```

Comment:
Do we have a plan to remove this function?  We could fix up the users when we update cockpitlib in subprojects.  Or are we keeping it in case we want to go back or have a longer transition period?

Reply:
There's exactly one caller in external projects, in c-machines. This was just a leftover, I added an extra commit here to kill it.